src/iccpd/src/mlacp_sync_prepare.c

```diff
@@ -0,0 +1,622 @@
+    #endif
+
+    return msg_len;
+}
+
+/*****************************************
+* Preprare Sync ARP-Info TLV
+*
+* ***************************************/
+int mlacp_prepare_for_arp_info(struct CSM* csm, char* buf, size_t max_buf_size, struct ARPMsg* arp_msg)
+{
+    struct mLACPARPInfoTLV* tlv = NULL;
+    size_t msg_len = 0;
+    size_t tlv_len = 0;
+    ICCHdr* icc_hdr = NULL;
+
+    if (!csm)
+        return -1;
+    if (!buf)
+        return -1;
+
+    tlv_len = sizeof(struct mLACPARPInfoTLV);
+
+    if ((msg_len = sizeof(ICCHdr) + tlv_len) > max_buf_size)
+        return -1;
+
+    /* ICC header */
+    memset(buf, 0, max_buf_size);
+    icc_hdr = (ICCHdr*)buf;
+    mlacp_fill_icc_header(csm, icc_hdr, msg_len);
+
+    /* Prepare for ARP information TLV */
+    tlv = (struct mLACPARPInfoTLV*)malloc(tlv_len);
+    memset(tlv, 0, tlv_len);
+    tlv->icc_parameter.u_bit = 0;
+    tlv->icc_parameter.f_bit = 0;
+    tlv->icc_parameter.type = htons(TLV_T_MLACP_ARP_INFO);
+    tlv->icc_parameter.len = htons(tlv_len - sizeof(ICCParameter));
+    tlv->type = arp_msg->op_type;
+    sprintf(tlv->ifname, "%s", arp_msg->ifname);
+    tlv->ipv4_addr = htonl(arp_msg->ipv4_addr);
+    memcpy(tlv->mac_addr, arp_msg->mac_addr, ETHER_ADDR_LEN);
+
+    /* Fill ARP Information TLV */
+    memcpy(&buf[sizeof(ICCHdr)], tlv, tlv_len);
+    free(tlv);
+
+    ICCPD_LOG_DEBUG(__FUNCTION__, "Prepare Msg if name %s  msg ifname %s  mac  =%02x:%02x:%02x:%02x:%02x:%02x ", tlv->ifname, arp_msg->ifname,  tlv->mac_addr[0], tlv->mac_addr[1], tlv->mac_addr[2],
+                    tlv->mac_addr[3], tlv->mac_addr[4], tlv->mac_addr[5]);
+    ICCPD_LOG_DEBUG(__FUNCTION__, "    IP Addr = %s ", show_ip_str( tlv->ipv4_addr));
+
+    return msg_len;
+}
+
+/*****************************************
+* Prprare Send portchannel info
+*
+* ***************************************/
+int mlacp_prepare_for_port_channel_info(struct CSM* csm, char* buf,
+                                        size_t max_buf_size,
+                                        struct LocalInterface* port_channel)
+{
+    struct System* sys = NULL;
+    ICCHdr* icc_hdr = NULL;
+    struct mLACPPortChannelInfoTLV* tlv = NULL;
+    size_t msg_len;
+    size_t tlv_len;
+    size_t name_len = MAX_L_PORT_NAME;
+    struct VLAN_ID* vlan_id = NULL;
+    int num_of_vlan_id = 0;
+
+    if (csm == NULL )
+        return -1;
+    if (buf == NULL )
+        return -1;
+    if (port_channel == NULL )
+        return -1;
+    if (port_channel->type == IF_T_PORT)
+        return -1;
+    if ((sys = system_get_instance()) == NULL )
+        return -1;
+
+    /* Calculate VLAN ID Length */
+    LIST_FOREACH(vlan_id, &(port_channel->vlan_list), port_next)
+    if (vlan_id != NULL)
+        num_of_vlan_id++;
+
+    tlv_len = sizeof(struct mLACPPortChannelInfoTLV) + sizeof(struct VLAN_ID) * num_of_vlan_id;
```

Comment:
tlv_len calculation is not correct. Only the size of the VLAN ID field (i.e. vid) should be counted instead of the entire struct VLAN_ID since only VLAN ID is sent to peer.

Reply:
> tlv_len calculation is not correct. Only the size of the VLAN ID field (i.e. vid) should be counted instead of the entire struct VLAN_ID since only VLAN ID is sent to peer.

@lisahnguyen, you are right, sizeof(struct VLAN_ID) must be sizeof(struct mLACPVLANData).